### ISSUE TITLE
Unexpected "thread blocked indefinitely in an MVar operation"

### ISSUE THREAD

**Issue by mtolly** [ghcjs/ghcjs#320]
I am getting what appear to be spurious `thread blocked indefinitely in an MVar operation` errors with GHCJS but not GHC. Here's a stripped-down example:

```haskell
module Main where

import Control.Concurrent
import Control.Concurrent.MVar

otherThread :: IO a -> IO a
otherThread f = do
  v <- newEmptyMVar
  _ <- forkIO $ f >>= putMVar v
  takeMVar v

main :: IO ()
main = do
  v <- otherThread $ otherThread $ do
    print "delaying" >> threadDelay 2000000 >> print "ok" >> return "success!"
  print v
```

`otherThread` just runs the action in another thread and passes the result back in an MVar. The key is that I have 2 `otherThread`s nested in `main`. If I have just one then there's no error. Also, the number given to `threadDelay` matters; when it's less than about 0.8 seconds for me, there's no error.

In browser console:

```
"delaying"
thread blocked indefinitely in an MVar operation
"ok"
```

In Node:

```
"delaying"
thread blocked indefinitely in an MVar operation
```

But compiled with normal GHC, I get the expected:

```
"delaying"
"ok"
"success!"
```

My original code is here: https://gist.github.com/mtolly/13e7e0217d350812580a
Basically I'm spawning threads to run actions in parallel, and then also spawning threads inside each action to set a time limit on it, so that's where the nested MVar locks came from.

My local `ghcjs` repo is at commit d3ddce090b22d528ddc928e1d3efe685e679f42b, so I'm assuming that's the version I have installed. If you think a newer version might fix it I can try upgrading.

**Comment by luite** [comment]
thanks for the testcase!

The deadlock check occurs as part of the heap reachability test (which also triggers finalizers and collects profiling info if enabled) and the rts sends an async exception to any thread blocked on an MVar that no other thread has references to. If your delay is too low, the program has finished before the first check occurs.

I've seen some similar issues from time to time, so I hope this testcase will make it easier to track down the problem.

**Comment by luite** [comment]
looks like live threads blocked on an mvar aren't marked properly, fix coming

**Comment by luite** [comment]
https://github.com/ghcjs/shims/commit/fbddb259767b72e673d0803938b9390a332befba

this should fix it

**Comment by mtolly** [comment]
Working great now, thanks much!
